### PR TITLE
feat: form defaults

### PR DESCRIPTION
**Pull request by yehjxraymond**
## Prepopulate default values in the form with defaults

Story: https://www.pivotaltracker.com/story/show/174033471

Previously `defaults` in the config file were ignored in the display of the form to the user. The intention of this PR is to allow the admin to specify what are some of the default values for fields that are meant to be editable to the user. One such example will be for the cover letter to specify a default title and description. 

To do this, we moved the configs into the form context to allow it to know the different form templates available and prepopulate the `formData` and `schema` object on creation of a new form.

## Additional Stuffs

In addition, the type for `Form` is renamed to `FormTemplate` to distinguish the different `FromTemplate`, `FromEntry` & `FormData`.

### CODE REVIEW

File: src/common/context/forms/index.tsx [hunk 1]
```diff
@@ -28,25 +30,32 @@ export const useFormsContext = (): FormsContext => useContext<FormsContext>(Form
 export const FormsContextProvider: FunctionComponent = ({ children }) => {
   const [activeFormIndex, setActiveFormIndex] = useState<number | undefined>(undefined);
   const [forms, setForms] = useState<FormEntry[]>([]);
+  const { config } = useConfigContext();
+
+  const currentForm = activeFormIndex === undefined ? undefined : forms[activeFormIndex];
```
Comment (Nebulis) [inline src/common/context/forms/index.tsx:35]: isn't `const currentForm = forms[activeFormIndex]` the same ?

File: src/common/context/forms/index.tsx [hunk 2]
```diff
@@ -28,25 +30,32 @@ export const useFormsContext = (): FormsContext => useContext<FormsContext>(Form
...
+  const currentFormData = currentForm ? currentForm.data : undefined;
```
Comment (Nebulis) [inline src/common/context/forms/index.tsx:36]: what version of TS you use ?

can eventually write it like `const currentFormData = currentForm?.data`

File: src/components/DynamicFormContainer/DynamicFormLayout.test.tsx [hunk 3]
```diff
@@ -96,21 +66,33 @@ const whenIsTransferableRecord = (): void => {
...
+    },
   });
 };
 
 const whenActiveFormIndexIsNotAvailable = (): void => {
-  mockUseConfigContext.mockReturnValue({ config: sampleConfig });
   mockUseFormsContext.mockReturnValue({
     activeFormIndex: undefined,
     setActiveFormIndex: mockSetActiveFormIndex,
+    currentFormTemplate: undefined,
```
Comment (Nebulis) [inline src/components/DynamicFormContainer/DynamicFormLayout.test.tsx:88]: why is this needed ?

File: src/test/fixtures/sample-local-config.json [hunk 4]
```diff
@@ -1,339 +1,344 @@
 {
```
Comment (Nebulis) [inline src/test/fixtures/sample-local-config.json:1]: can help me to know the diff ?

File: src/common/context/forms/index.tsx [hunk 5]
```diff
@@ -28,25 +30,32 @@ export const useFormsContext = (): FormsContext => useContext<FormsContext>(Form
...
+
     setForms([
       ...forms,
       {
         templateIndex,
-        data: {} as any, // eslint-disable-line @typescript-eslint/no-explicit-any
+        data: {
+          formData: newFormTemplate?.defaults || {},
+          schema: newFormTemplate?.schema,
+        } as any, // eslint-disable-line @typescript-eslint/no-explicit-any
```
Comment (Nebulis) [inline src/common/context/forms/index.tsx:51]: any is wrong here right ?

it's an object with properties `formData` and `schema` that are any

File: src/common/context/forms/forms.test.tsx [hunk 6]
```diff
@@ -1,12 +1,24 @@
...
 
 const wrapper: FunctionComponent = ({ children }) => (
   <FormsContextProvider>{children}</FormsContextProvider>
 );
 
 describe("useFormsContext", () => {
+  beforeAll(() => {
+    mockUseConfigContext.mockReturnValue({
+      config: {
+        forms: [{ schema: "FORM_1_SCHEMA" }, { defaults: { foo: "bar" }, schema: "FORM_2_SCHEMA" }],
```
Comment (Nebulis) [inline src/common/context/forms/forms.test.tsx:18]: suggestion to rename: defaults => defaultValues

defaults is a bit generic and might become eventually problematic to understand if we need to introduce any other default related parameter

File: src/common/context/forms/index.tsx (hunk 1 above)
Comment (yehjxraymond) [reply src/common/context/forms/index.tsx:35]: hmm apparently ts throws when there is undef in the [].

image

Comment (yehjxraymond) [reply src/common/context/forms/index.tsx:35]: The alternative is then `activeFormIndex ? forms[activeFormIndex] : undefined`, but when the index = 0, then it will be undefined.

File: src/components/DynamicFormContainer/DynamicFormLayout.test.tsx (hunk 3 above)
Comment (yehjxraymond) [reply src/components/DynamicFormContainer/DynamicFormLayout.test.tsx:88]: its not needed, was just being more explicit

File: src/test/fixtures/sample-local-config.json (hunk 4 above)
Comment (yehjxraymond) [reply src/test/fixtures/sample-local-config.json:1]: network: local

Comment (yehjxraymond) [reply src/test/fixtures/sample-local-config.json:1]: its for use on local ganache network

File: src/common/context/forms/index.tsx (hunk 5 above)
Comment (yehjxraymond) [reply src/common/context/forms/index.tsx:51]: yep right, removed

File: src/common/context/forms/forms.test.tsx (hunk 6 above)
Comment (yehjxraymond) [reply src/common/context/forms/forms.test.tsx:18]: hmm i dont necessary agree with this one. I'm indifferent to either, just have to change all the config file to use it. If someone wants to do it out of this PR I wont object.

image
